models/staircase.py

```diff
@@ -0,0 +1,374 @@
+            return None
+        return h[:, start_idx:end_idx, :]
+
+    def forward(self, x, h_prev, target=None, **kargs):
+        # input h B x M
+        # assume h_prev [B, staircase_size], and will init with 0s
+        # create output placeholder
+        # output [B, mem_size, hidden_size]
+        hid_after_embed = self.in_emb(x)  # B x M x H
+        if self.args.emb_drop > 0:
+            hid_after_embed = self.emb_dropout(hid_after_embed)
+        # no cache between forwards
+
+        output = hid_after_embed.new_zeros(
+            (hid_after_embed.size(0), self.mem_size, self.hidden_size)
+        )
+        start_idx = 0
+        # generate scheduling for staircases, randomness for training purpose
+        total_steps = (
+            self.mem_size + self.staircase_size
+        ) // self.fix_staircase_size_forward - 1
+        prev_output = None
+        for step_idx in range(total_steps):
+            end_idx = start_idx + self.fix_staircase_size_forward
+            cache = self.get_cache(h_prev, step_idx)
+            new_tokens = self.get_new_tokens(
+                hid_after_embed, start_idx, end_idx)
+            # should put into cache:
+            cache_id = step_idx - (
+                total_steps
+                - (self.staircase_size // self.fix_staircase_size_forward)
+                + 1
+            )
+            # consumed all forward steps
+            # when should we move prev_output forwards
+            if step_idx >= self.staircase_size // self.fix_staircase_size_forward:
+                prev_output = prev_output[:,
+                                          self.fix_staircase_size_forward:, :]
+            # input to the first layer of the model
+            if self.variant == VARIANT.VARIANT_ONE:
+                context = self.assemble_context(
+                    cache[0], prev_output, new_tokens)
+                h = self.assemble_query(prev_output, new_tokens)
+                assert context.size(1) <= self.staircase_size
+                # forward into layers
+                cache_for_next = []
+                for layer in range(self.args.nlayers):
+                    # only forward size get updated from the context as well
+                    h = self.transformer.get_layer(layer)(h, context)
+                    if h.size(1) == context.size(1):
+                        # self-attention
+                        context = h
+                    elif layer + 1 < self.args.nlayers:
+                        context = torch.cat([cache[layer + 1], h], dim=1)
+                    # put into cache
+                    if cache_id >= 0:
+                        cache_for_next.append(h)
+                # the output from the last layer
+                prev_output = h
+                # put into cache
+                if len(cache_for_next) > 0:
+                    h_prev[cache_id] = cache_for_next
+            else:
+                raise RuntimeError("Variant 3 and 4 is in another file!")
```

Comment:
Are you planning to add variant 2, 3, and 4? If not, maybe you could just remove the variant arg, enum, and property!